[#277] Use timbre's sentry appender to log/capture the Exception

iperdomo opened the pull request:
* Setup the sentry endpoint in wiremock to avoid a 'warning' in the
  logs when running the tests

iperdomo committed:
[#277] Use timbre's sentry appender to log/capture the Exception

* Setup the sentry endpoint in wiremock to avoid a 'warning' in the
  logs when running the tests

dlebrero commented on backend/test/akvo/flow_services/end_to_end_test.clj:
```diff
@@ -203,6 +203,7 @@
     (mock-flow-report-api flow-report-id)
     (mock-gae survey-id)
     (test-util/mock-mailjet)
+    (test-util/mock-sentry)
```
I now remember that I did not setup the mock sentry because if we do, if the test fails due to an exception, this exception is **just** going to be sent to the Sentry mock, so basically it is going to disappear in the WireMock ether, leaving the Dev clueless about what happened.

Maybe add an assert in this particular test to check that there were no exceptions sent to Sentry (and log the actual exception sent to Sentry to help the dev figure out the issue) or do not mock Sentry for this particular test.

dlebrero commented on backend/src/akvo/flow_services/error.clj:
```diff
@@ -25,7 +26,7 @@
      ~body
      (catch Exception e#
-       (.printStackTrace e#)
+       (log/error e#)
```
It feels weird that a function that tries to make the error handling more FP has a side effect on it. 

Apart, if there is an error when calling Flow for the notification, like in https://github.com/akvo/akvo-flow-services/blob/500c8b077013a3ded88dbe95e295569e223fe1fc/backend/src/akvo/flow_services/scheduler.clj#L116 then we are going to end up logging the exception twice, once is the log line of this change and another one due to https://github.com/akvo/akvo-flow-services/blob/cd2b30caaf34e2db458e46fbb1aaf2ec30372c65/backend/src/akvo/flow_services/core.clj#L130

A suggestion would to log the error on the "non-pure-FP" side at
https://github.com/akvo/akvo-flow-services/blob/500c8b077013a3ded88dbe95e295569e223fe1fc/backend/src/akvo/flow_services/scheduler.clj#L117
by changing that `(when (e/if-ok...` to `(e/fmap...` and use `timbre/error` in the fn-error param

iperdomo committed:
[#277] Apply changes based on feedback from @dlebrero

When using fmap I was not able to use directly log/error from timbre.
Not sure if is because is a macro or not. I had to wrap the call in an
anonymous function.

dlebrero commented on backend/src/akvo/flow_services/scheduler.clj:
```diff
@@ -114,8 +114,9 @@
-    (when (e/ok? report)
-      (gdpr-email job-data))))
+    (e/fmap report
+            #(log/error (-> % ::e/error :cause))
```
I think `fmap` unwraps the `::e/error`

dlebrero commented on backend/test/akvo/flow_services/end_to_end_test.clj:
```diff
@@ -217,13 +217,15 @@
       (test-util/check-report-file-headers report-result "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
-      (is (= ["IN_PROGRESS" "FINISHED_SUCCESS"] (final-report-state-in-flow flow-report-id))))))
+      (is (= ["IN_PROGRESS" "FINISHED_SUCCESS"] (final-report-state-in-flow flow-report-id)))
+      (is (zero? (sentry-alerts-count))))))
```
Can we assert that is always `zero?`? Does the wire mock counter resets when we do the setup?

iperdomo replied:
There is a _fixture_ that call wiremock reset for each test case, so we know that this should be zero.

dlebrero replied:
Ta!

iperdomo committed:
[#277] error/fmap unwraps the error already

We're going to log as error map similar to:

    {:cause (Exception. "Some error")}
